### DISCUSSION TITLE
BooleanBuilder with multiple AND conditions

### DISCUSSION THREAD

**Discussion by dazerek** [querydsl/querydsl#2936] [Q&A]
Hi,
I'm using QueryDSL with MongoDB and Spring Boot.
```
<querydsl.version>4.3.1</querydsl.version>
<mysema.maven.version>1.1.3</mysema.maven.version>
```
I'm building complex search using multiple AND conditions, so I've created Predicate Builder, which makes below BooleanBuilder:

```
public static Predicate toPredicate(final MetaDataSearch metaDataSearch) {
        final QMetaData qMetaData = QMetaData.metaData;
        BooleanBuilder booleanBuilder = new BooleanBuilder();
        metaDataSearch.getCreatedDateFrom().ifPresent(dateFrom -> {
            booleanBuilder.and(qMetaData.createdAtDate.after(dateFrom));
        });
        metaDataSearch.getCreatedDateTo().ifPresent(dateTo -> {
            booleanBuilder.and(qMetaData.createdAtDate.before(dateTo));
        });
        metaDataSearch.getFileTypeCode().ifPresent(fileTypeCode -> {
            booleanBuilder.and(qMetaData.fileTypeCode.equalsIgnoreCase(fileTypeCode));
        });
        metaDataSearch.getFileName().ifPresent(fileName -> {
            booleanBuilder.and(qMetaData.fileName.equalsIgnoreCase(fileName));
        });
        metaDataSearch.getUserId().ifPresent(userId -> {
            booleanBuilder.and(qMetaData.userId.eq(userId.toString()));
        });
        metaDataSearch.getAdditionalProperties().ifPresent(
            additionalProperties ->
                additionalProperties.forEach((key, value) ->
                    booleanBuilder.and(qMetaData.additionalProperties.contains(key, value))));
        return booleanBuilder;
    }
```

The issue is that executing AND method multiple times on BooleanBuilder ends up with wrong query:

`find using query: { "$and" : [{ "createdAtDate" : { "$gt" : { "$date" : "2020-08-19T09:45:14Z"}}, "additionalProperties.clientID" : 123212}, { "createdAtDate" : { "$lt" : { "$date" : "2020-08-19T10:26:14Z"}}}]}`

However executing OR method ends up with correct query:
`query: { "$or" : [{ "createdAtDate" : { "$gt" : { "$date" : "2020-08-19T09:48:56Z"}}}, { "createdAtDate" : { "$lt" : { "$date" : "2020-08-19T10:29:56Z"}}}, { "additionalProperties.clientID" : 123212}]}`

I also tested methods:

http://www.querydsl.com/static/querydsl/latest/apidocs/com/querydsl/core/types/dsl/Expressions.html#allOf-com.querydsl.core.types.dsl.BooleanExpression...-
http://www.querydsl.com/static/querydsl/latest/apidocs/com/querydsl/core/types/ExpressionUtils.html#allOf-com.querydsl.core.types.Predicate...-

None of them seems to work properly with multiple BooleanExpressions/Predicates combined with AND condition.

**Comment by jwgmeligmeyling** [comment]
Please retest against a modern version of QueryDSL. We're not using the `mysema` groupId for years now, so you really must be on an old version. Hope that helps!

**Comment by dazerek** [comment]
> Please retest against a modern version of QueryDSL. We're not using the `mysema` groupId for years now, so you really must be on an old version. Hope that helps!

Thanks for reply, followed latest docs:
http://www.querydsl.com/static/querydsl/latest/reference/html/ch02s07.html

Could u provide up-to-date documentation?

**Comment by dazerek** [comment]
> Please retest against a modern version of QueryDSL. We're not using the `mysema` groupId for years now, so you really must be on an old version. Hope that helps!

Switched from mysema into different annotationProcessor:
```
 <plugin>
        <groupId>org.bsc.maven</groupId>
        <artifactId>maven-processor-plugin</artifactId>
        <version>4.2</version>
        <executions>
          <execution>
            <id>process</id>
            <goals>
              <goal>process</goal>
            </goals>
            <phase>generate-sources</phase>
            <configuration>
              <processors>
                <processor>com.querydsl.apt.morphia.MorphiaAnnotationProcessor</processor>
              </processors>
              <outputDirectory>${project.build.directory}/generated-sources/annotations
              </outputDirectory>
            </configuration>
          </execution>
        </executions>
      </plugin>
```


 which didn't change anything as my QMetaData is the same:
```
/**
 * QMetaData is a Querydsl query type for MetaData
 */
@Generated("com.querydsl.codegen.EntitySerializer")
public class QMetaData extends EntityPathBase<MetaData> {

    private static final long serialVersionUID = -547961607L;

    public static final QMetaData metaData = new QMetaData("metaData");

    public final MapPath<String, Object, SimplePath<Object>> additionalProperties = this.<String, Object, SimplePath<Object>>createMap("additionalProperties", String.class, Object.class, SimplePath.class);

    public final StringPath comment = createString("comment");

    public final DateTimePath<java.time.LocalDateTime> createdAtDate = createDateTime("createdAtDate", java.time.LocalDateTime.class);

    public final StringPath fileName = createString("fileName");

    public final StringPath fileStorageId = createString("fileStorageId");

    public final StringPath fileTypeCode = createString("fileTypeCode");

    public final StringPath id = createString("id");

    public final DateTimePath<java.time.LocalDateTime> modifyAtDate = createDateTime("modifyAtDate", java.time.LocalDateTime.class);

    public final EnumPath<MetaDataStatus> status = createEnum("status", MetaDataStatus.class);

    public final StringPath userId = createString("userId");

    public QMetaData(String variable) {
        super(MetaData.class, forVariable(variable));
    }

    public QMetaData(Path<? extends MetaData> path) {
        super(path.getType(), path.getMetadata());
    }

    public QMetaData(PathMetadata metadata) {
        super(MetaData.class, metadata);
    }

}
```

However re-tested and faced same issue, BooleanBuilder comes from com.querydsl.core which is set by:
`<querydsl.version>4.3.1</querydsl.version>`
and below deps:

```
   <dependency>
      <groupId>com.querydsl</groupId>
      <artifactId>querydsl-mongodb</artifactId>
      <version>${querydsl.version}</version>
    </dependency>
    <dependency>
      <groupId>com.querydsl</groupId>
      <artifactId>querydsl-apt</artifactId>
      <version>${querydsl.version}</version>
    </dependency>
```

**Comment by jwgmeligmeyling** [comment]
The documentation is indeed outdated (https://github.com/querydsl/querydsl/issues/2631). Up until recently, the contributors were still struggling to get access to the host and DNS (https://querydsl.slack.com/archives/C06KLG1BP/p1590583190037900). Although development on `querydsl-core`, `querydsl-jpa` and `querydsl-sql` is very active at the moment, there's little interest among the existing contributors in `querydsl-mongodb` (https://github.com/querydsl/querydsl/issues/2459#issuecomment-567237655). Contributors are welcome!

That said, this seems to be a problem in how expressions are converted into a MongoDB query. That translation happens in the `MongodbSerializer`. For `AND` and `ON` operations specifically, it is handled in lines https://github.com/querydsl/querydsl/blob/master/querydsl-mongodb/src/main/java/com/querydsl/mongodb/MongodbSerializer.java#L109-L139 . If it can be done, a fix should likely be made there.
